pubsub/google/cloud/pubsub_v1/subscriber/policy/base.py

```diff
@@ -0,0 +1,392 @@
+        return request
+
+    def lease(self, ack_id, byte_size):
+        """Add the given ack ID to lease management.
+
+        Args:
+            ack_id (str): The ack ID.
+            byte_size (int): The size of the PubSub message, in bytes.
+        """
+        # Add the ack ID to the set of managed ack IDs, and increment
+        # the size counter.
+        if ack_id not in self.managed_ack_ids:
+            self.managed_ack_ids.add(ack_id)
+            self._bytes += byte_size
+
+        # Sanity check: Do we have too many things in our inventory?
+        # If we do, we need to stop the stream.
+        if self._load >= 1.0:
+            self._paused = True
+            self.close()
+
+    def maintain_leases(self):
+        """Maintain all of the leases being managed by the policy.
+
+        This method modifies the ack deadline for all of the managed
+        ack IDs, then waits for most of that time (but with jitter), and
+        then calls itself.
+
+        .. warning::
+            This method blocks, and generally should be run in a separate
+            thread or process.
+
+            Additionally, you should not have to call this method yourself,
+            unless you are implementing your own policy. If you are
+            implementing your own policy, you _should_ call this method
+            in an appropriate form of subprocess.
+        """
+        # Sanity check: Should this infinitely-recursive loop quit?
+        if not self._consumer.active:
+            return
+
+        # Determine the appropriate duration for the lease.
+        # This is based off of how long previous messages have taken to ack,
+        # with a sensible default and within the ranges allowed by Pub/Sub.
+        p99 = self.histogram.percentile(99)
+        logger.debug('The current p99 value is %d seconds.' % p99)
+
+        # Create a streaming pull request.
+        # We do not actually call `modify_ack_deadline` over and over because
+        # it is more efficient to make a single request.
+        ack_ids = list(self.managed_ack_ids)
+        logger.debug('Renewing lease for %d ack IDs.' % len(ack_ids))
+        if len(ack_ids) > 0 and self._consumer.active:
+            request = types.StreamingPullRequest(
+                modify_deadline_ack_ids=ack_ids,
+                modify_deadline_seconds=[p99] * len(ack_ids),
+            )
+            self._consumer.send_request(request)
+
+        # Now wait an appropriate period of time and do this again.
+        #
+        # We determine the appropriate period of time based on a random
+        # period between 0 seconds and 90% of the lease. This use of
+        # jitter (http://bit.ly/2s2ekL7) helps decrease contention in cases
+        # where there are many clients.
+        snooze = random.uniform(0.0, p99 * 0.9)
+        logger.debug('Snoozing lease management for %f seconds.' % snooze)
+        time.sleep(snooze)
+        self.maintain_leases()
```

Comment:
this will eventually hit the recursion limit. Prefer splitting into a `update_leases` single-shot and a `maintain leases` that while loops, or just add a while loop to this one.